docs/features/source-generators.md

```diff
@@ -0,0 +1,291 @@
+every generator on every keystroke, and still maintain an acceptable level of performance 
+within the IDE.
+
+#### Progressive complexity opt-in
+
+It is expected instead that source generators would work on an 'opt-in' approach to IDE 
+enablement. 
+
+By default, a generator implementing only `ISourceGenerator` would see no IDE integration
+and only be correct at build time. Based on conversations with 1st party customers, 
+there are several cases where this would be enough.
+
+However, for scenarios such as code first gRPC, and in particular Razor and Blazor, 
+the IDE will need to be able to generate code on-they-fly as those file types are 
+edited and reflect the changes back to other files in the IDE in near real-time. 
+
+The proposal is to have a set of advanced interfaces that can be optionally implemented,
+that would allow the IDE to query the generator to decide what needs to be run in the case
+of any particular edit.
+
+For example an extension that would cause generation to run after saving a third party
+file might look something like: 
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface ITriggeredByAdditionalFileSavedGenerator : ISourceGenerator
+    {
+        ImmutableArray<string> SupportedAdditionalFileExtensions { get; }
+    }
+}
+```
+It is expected that there will be various levels of opt in, that can be added to a generator
+in order to achieve the specific level of performance required of it. 
+
+What these exact APIs will look like remains an open question, and it's expected that we will 
+need to prototype some real-world generators before knowing what their precise shape will be.
+
+### Output files
+
+It it desirable that the generated source texts be available for inspection after generation,
+either as part of creating a generator or seeing what code was generated by a third party
+generator.
+
+By default, generated texts will be persisted to a `GeneratedFiles/{GeneratorAssemblyName}` 
+sub-folder within `CommandLineArguments.OutputDirectory`. The `fileNameHint` from 
+`SourceGeneratorContext.AddSource` will be used to create a unique name, with appropriate
+collision renaming applied if required. For instance, on Windows a call to 
+`AddSource("MyCode", ...);` from `MyGenerator.dll` for a C# project might be 
+persisted as `obj/debug/GeneratedFiles/MyGenerator.dll/MyCode.cs`.
+
+File output is not required for the correct function of either command line or IDE based 
+generation, and can be completely disabled, if required. The IDE will work on in-memory 
+copies of the generated source texts (for 'Find all references', breakpoints etc.) and 
+periodically flush any changes to disk. 
+
+To support the use case where a user wishes to generate the source text, then commit 
+the generated files to source control, we will allow changing the location of the 
+generated files via an appropriate command line switch, and matching MSBuild property 
```

Comment:
Something to consider: if we can figure out the output names of the generator, and have a list of output items with the same name, we can decide whether or not they are up-to-date based on file stamps.

We might want to try to get this in V1, since it may be difficult to add after-the-fact.

Reply:
Added to the doc to track

---
In reply to: [355151636](https://github.com/dotnet/roslyn/pull/40162#discussion_r355151636) [](ancestors = 355151636)